### PR TITLE
[1866] Germany & Italy

### PR DESCRIPTION
**Pull request by magnusfalt**
- Public companies convert into a 10 share company from phase 3. When converting, the director can buy up to 60% of the shares.
- Germany and Italy national conversion from phase 3, must at phase 5
- Forced Germany or Italy without any president need will "run", it tiles in country move one step to right, otherwise one step to left.

### CODE REVIEW

File: lib/engine/game/g_1866/game.rb [hunk 1]
```diff
@@ -61,6 +64,12 @@ class Game < Game::Base
                                               par_2: 'Brown phase (5/6) par',
                                               par_3: 'Gray phase (8/10) par').freeze
 
+        STATUS_TEXT = Base::STATUS_TEXT.merge(
+          'can_convert_public' => ['Convert Public', 'Public companies can convert from 5 shares to 10 shares.'],
```
Comment (tobymao) [inline lib/engine/game/g_1866/game.rb:68]: we should use corporation throughout to be consistent with 18xx.games

File: lib/engine/game/g_1866/game.rb [hunk 2]
```diff
@@ -950,20 +980,100 @@ def upgrades_to?(from, to, _special = false, selected_company: nil)
...
         end
 
+        def can_par_share_price?(share_price, corporation)
+          return (share_price.corporations.empty? || share_price.price == self.class::MAX_PAR_VALUE) unless corporation
+
+          share_price.corporations.none? { |c| c.type != :stock_turn_corporation } ||
+            share_price.price == self.class::MAX_PAR_VALUE
+        end
+
+        def convert_public_corporation?
```
Comment (tobymao) [inline lib/engine/game/g_1866/game.rb:990]: what's a public corporation, is there a private corporation?

File: lib/engine/game/g_1866/step/convert.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            if entity != current_entity || !@game.public_corporation?(entity) || !@game.convert_public_corporation? ||
+              entity.type != :public_5
+              return []
+            end
+
+            ACTIONS
+          end
+
+          def choice_name
+            "Convert #{current_entity.id} from 5 share to 10 share company"
```
Comment (tobymao) [inline lib/engine/game/g_1866/step/convert.rb:22]: corporation

File: lib/engine/game/g_1866/step/convert.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            # Create 5 new shares
+            shares = Array.new(5) { |i| Share.new(entity, percent: 10, index: i + 4) }
+            shares.each do |share|
+              entity.share_holders[entity] += share.percent
+              entity.shares_by_corporation[entity] << share
+            end
+            entity.type = :public_10
+
+            # Buy the shares
+            @log << "#{entity.name} converts to a 10-share company"
```
Comment (tobymao) [inline lib/engine/game/g_1866/step/convert.rb:71]: corporation

File: lib/engine/game/g_1866/game.rb (hunk 2 above)
Comment (magnusfalt) [reply lib/engine/game/g_1866/game.rb:990]: In rules there are many different types of companies, but the 2 with private and public is
private company = company in 18xx.games
public company = corporation

Ive just looked at the rules to much and copied :)

**Comment by magnusfalt** [comment]
Cleaned up the code so we get consistency with the rest of 18xx.games

**Review by tobymao — APPROVED** [review]

**Pull request merged**